File: Python/ConvertBSTToGreaterTree/ConvertBSTToGreaterTree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    prevSum = 0
    # (122ms)
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def traverse(root):
            if not root:
                return
            traverse(root.right)
            root.val += self.prevSum
            self.prevSum = root.val
            traverse(root.left)
        traverse(root)
        return root

File: JavaScript/ConvertBSTToGreaterTree/ConvertBSTToGreaterTree.js
/**
 * Created by Phoenix-Z on 2017/4/21.
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * (215ms)
 * @param {TreeNode} root
 * @return {TreeNode}
 */

var convertBST = function(root) {
    var traverse = function (root) {
        if(root === null)
            return;
        traverse(root.right);
        root.val += prevSum;
        prevSum = root.val;
        traverse(root.left);
    };
    var prevSum = 0;
    traverse(root);
    return root;
};

